internal/app/ntpd/pkg/ntp/ntp.go

```diff
@@ -5,93 +5,110 @@
-const (
-	MaxPoll = 1000
-	MinPoll = 20
+	"github.com/hashicorp/go-multierror"
 )
 
 // NTP contains a server address
-// and the most recent response from a query
 type NTP struct {
-	Server   string
-	Response *ntp.Response
+	Server  string
+	MinPoll time.Duration
+	MaxPoll time.Duration
+	Retry   int
 }
 
 // NewNTPClient instantiates a new ntp client for the
 // specified server
-func NewNTPClient(server string) *NTP {
-	return &NTP{Server: server}
+func NewNTPClient(opts ...Option) (*NTP, error) {
+	ntp := defaultOptions()
+
+	var result *multierror.Error
+	for _, setter := range opts {
+		result = multierror.Append(setter(ntp))
+	}
+
+	return ntp, result.ErrorOrNil()
 }
 
 // Daemon runs the control loop for query and set time
 // We dont ever want the daemon to stop, so we only log
 // errors
 func (n *NTP) Daemon() (err error) {
-	rando := rand.New(rand.NewSource(time.Now().UnixNano()))
-	ticker := time.NewTicker(time.Duration(rando.Intn(MaxPoll)+MinPoll) * time.Second)
 
 	// Do an initial hard set of time to ensure clock skew isnt too far off
 	var resp *ntp.Response
-	resp, err = n.Query()
-	if err != nil {
+	if resp, err = n.Query(); err != nil {
 		log.Printf("error querying %s for time, %s", n.Server, err)
 		return err
 	}
-	n.Response = resp
 
-	if err = n.SetTime(resp); err != nil {
+	if err = setTime(resp.Time); err != nil {
 		return err
 	}
 
+	var randSleep time.Duration
 	for {
-		<-ticker.C
-		// Set some variance with how frequently we poll ntp servers
-		resp, err = n.Query()
-		if err != nil {
+		// Set some variance with how frequently we poll ntp servers.
+		// This is based on rand(MaxPoll) + MinPoll so we wait at least
+		// MinPoll.
+		randSleep = time.Duration(rand.Intn(int(n.MaxPoll.Seconds()))) * time.Second
+		time.Sleep(randSleep + n.MinPoll)
+
+		if resp, err = n.Query(); err != nil {
 			// As long as we set initial time, we'll treat
 			// subsequent errors as nonfatal
 			log.Printf("error querying %s for time, %s", n.Server, err)
 			continue
 		}
-		n.Response = resp
 
-		if err = n.SetTime(resp); err != nil {
+		if err = adjustTime(resp.ClockOffset); err != nil {
 			log.Printf("failed to set time, %s", err)
 			continue
 		}
-		ticker = time.NewTicker(time.Duration(rando.Intn(MaxPoll)+MinPoll) * time.Second)
 	}
 }
 
-// Query polls the ntp server to get back a response
-// and saves it for later use
+// Query polls the ntp server and verifies a successful response.
 func (n *NTP) Query() (*ntp.Response, error) {
-	return ntp.Query(n.Server)
-}
 
-// SetTime sets the system time based on the query response
-func (n *NTP) SetTime(resp *ntp.Response) error {
-	// Not sure if this is the right thing to do
-	if resp == nil {
-		return errors.New("not a valid ntp response")
+	for i := 0; i < n.Retry; i++ {
+		resp, err := ntp.Query(n.Server)
+		if err != nil {
+			time.Sleep(time.Duration(i) * n.MinPoll)
+			continue
+		}
+
+		if err := resp.Validate(); err != nil {
```

Comment:
This func handles the juicy details,

https://github.com/beevik/ntp/blob/master/ntp.go#L248-L252